### PR TITLE
[Whisper Tokenizer] Make more user-friendly

### LINKED ISSUE #19864: Whisper's Tokenizer encoding function is not user-friendly
### System Info

- `transformers` version: 4.23.0.dev0
- Platform: Linux-5.3.0-64-generic-x86_64-with-glibc2.17
- Python version: 3.8.12
- Huggingface_hub version: 0.10.0
- PyTorch version (GPU?): 1.12.1+cu102 (True)
- Tensorflow version (GPU?): 2.9.1 (True)
- Flax version (CPU?/GPU?/TPU?): 0.4.1 (cpu)
- Jax version: 0.3.14
- JaxLib version: 0.3.14
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

Text encoded by Whisper should by default have a EOS token in the end (just like any sequence-to-sequence transformer model) and also include the correct other tokens if necessary. Also it should **not** return the `attention_mask` as Whisper never uses an `attention_mask`.

**Note**: Encoding of text tokens is only needed for fine-tuning the model, not for inference, so this bug/feature request is not relevant for inference.

Currently when doing:

```python
from transformers import WhisperTokenizer
tokenizer = WhisperTokenizer.from_pretrained("openai/whisper-small")

tokenizer("hey")
```
One gets:
```
{'input_ids': [17230], 'attention_mask': [1]}
```

There are multiple problems with this:
- The EOS token is not appended, but whisper **always** needs an EOS for training
- There should be the following config parameters for the whisper tokenizer:
  - Set a language id so that the tokenizer automatically adds the lang id prefix
  - Set a "use_timestamps" true/false flag to the tokenizer that decides whether the "notimesteps" token should be added or not
  - Remove this line: https://github.com/huggingface/transformers/blob/371337a95b5d82cc9376c2595ed2022a5eb2ee6e/src/transformers/models/whisper/tokenization_whisper.py#L119 as this is not relevant for whisper and a copy-paste from GPT2 -> whisper should never set the EOS token as the BOS token IMO (@sanchit-gandhi we did this for our paper, but this was more a lucky bug than a solid case)
- Whisper should **not** return the `attention_mask`

=> Let's try to make whisper as user-friendly as possible for fine-tuning. Say I'd like to fine-tune on a multi-lingual language

If I remember correctly the format for the **labels** to fine-tune whisper on multi-lingual is:

```
<lang-id><|notimestamps|> ... text ... <eos>
```

@ArthurZucker or is it first `<|notimestamps|>` and then `<lang-id>` and then <eos> ? 

The `decoder_input_ids` should then just add `<|startoftranscript>` and remove `<eos>` which happens automatically in the shift function.

Now I should be able to get this encoding behavior by default when doing:

```
from transformers import WhisperTokenizer
tokenizer = WhisperTokenizer.from_pretrained("openai/whisper-small", lang_id="dv", predict_timestamps=False)

tokenizer("hey")
```

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

from transformers import WhisperTokenizer
tokenizer = WhisperTokenizer.from_pretrained("openai/whisper-small", lang_id="dv", predict_timestamps=False)

tokenizer("hey")

### Expected behavior

```
<lang-id><|notimestamps|> ... text ... <eos>
```

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
# What does this PR do?

Fixes #19864.

In summary, the Whisper tokenizer is modified to prepend several tokens to the start-of-sequence:
- BOS token id (`<|startoftranscript|>`) -> consistent with other sequence-to-sequence models such as _BART_.
- Language token id (e.g. `<|es|>` for Spanish) -> set only when the tokenizer is instantiated with argument `language=X`. Otherwise omitted.
- Task token id (e.g. `<|translate|>` for speech translation) -> set only when the tokenizer is instantiated with argument `task=Y`. Otherwise omitted.
- No time stamps id (`<|notimestamps|>`)  ->set only when the tokenizer is instantiated with argument  `predict_timestamps=False`. For `predict_timestamps=True`, it is omitted.

In addition, it is modified to always append the end-of-sequence token to the end of the label sequence (`<|endoftext|>`).

The updated tokenizer behaves as follows:
```python
from transformers import WhisperTokenizer
tokenizer = WhisperTokenizer.from_pretrained("openai/whisper-tiny", language="english", task="transcribe", predict_timestamps=False)

input_ids = tokenizer("hey").input_ids

text_with_special = tokenizer.decode(input_ids, skip_special_tokens=False)
text = tokenizer.decode(input_ids, skip_special_tokens=True)

print("Input ids :", input_ids)
print("Text w/ special :", text_with_special)
print("Text :", text)
```
**Print Output:**
```
Input ids : [50258, 50259, 50359, 50363, 17230, 50257]
Text w/ special : <|startoftranscript|><|en|><|transcribe|><|notimestamps|>hey<|endoftext|>
Text : hey
```

The attention mask functionality of the Whisper tokenizer **is** retained (_c.f._ https://github.com/huggingface/transformers/issues/19864#issuecomment-1291799687).

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: tests/models/whisper/test_tokenization_whisper.py [hunk 1]
```diff
@@ -138,35 +137,42 @@ def test_tokenizer_equivalence(self):
...
         # fmt: on
 
-        self.assertListEqual(gpt2_tokens, EXPECTED_ENG)
+        self.assertListEqual(monolingual_tokens, EXPECTED_ENG)
         self.assertListEqual(multilingual_tokens, EXPECTED_MULTI)
 
     def test_tokenizer_special(self):
-        multilingual_tokenizer = WhisperTokenizer.from_pretrained("openai/whisper-tiny.en")
-        text = "<|startoftranscript|>Hey! How are you feeling? J'ai l'impression que 郷さん est prêt<|endoftext|>"
+        multilingual_tokenizer = WhisperTokenizer.from_pretrained(
```
Comment (sanchit-gandhi) [inline tests/models/whisper/test_tokenization_whisper.py:144]: Refactored to use a multilingual tokenizer.

File: src/transformers/models/whisper/tokenization_whisper.py [hunk 2]
```diff
@@ -89,9 +89,130 @@ def get_pairs(word):
     return pairs
 
 
+LANGUAGES = {
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/tokenization_whisper.py:92]: Currently using `language` (e.g. `spanish`) instead of `lang_id` (`es`) -> this is how the original Whisper model does it. If there's a preference for `lang_id` I'm happy to switch!

File: src/transformers/models/whisper/tokenization_whisper.py [hunk 3]
```diff
@@ -231,27 +361,57 @@ def bpe(self, token):
         self.cache[token] = word
         return word
 
-    # Copied from transformers.models.gpt2.tokenization_gpt2.GPT2Tokenizer.build_inputs_with_special_tokens with GPT2 -> Whisper
-    def build_inputs_with_special_tokens(self, token_ids_0, token_ids_1=None):
-        if self.add_bos_token:
-            bos_token_ids = [self.bos_token_id]
-        else:
-            bos_token_ids = []
+    def set_bos_sequence(self, language=None, task=None, predict_timestamps=False):
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/whisper/tokenization_whisper.py:364]: Also wondering if we should maybe do 
`self.language = language if language is not None`, that you can also set the prefix tokens by calling this function once. WDYT? Or we can have a real `set_prefix_tokens` somewhere else.

**Review by ArthurZucker — COMMENTED** [review]
Awesome work here! A few nits here an there but thanks a lot
Do you think we can also update the doc or write somewhere on the model card that in order to train the model, just pass the language and run it.

File: src/transformers/models/whisper/tokenization_whisper.py [hunk 4]
```diff
@@ -231,27 +361,57 @@ def bpe(self, token):
...
+        all_special_ids = self.all_special_ids
+        bos_token_id = all_special_ids[-106]
+        translate_token_id = all_special_ids[-6]
+        transcribe_token_id = all_special_ids[-5]
+        notimestamps_token_id = all_special_ids[-1]
```
[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply src/transformers/models/whisper/tokenization_whisper.py:369]: I see your point! The only downside of using `all_special_tokens` is that we'll have to do an extra tokenisation step of converting from tokens -> id's in this method:
```python
bos_token = all_special_tokens[-106]
translate_token = all_special_tokens[-6]
...
# get prefix tokens (bos_token, lang_token, task_token, notimestamps_token)
...
prefix_ids = self.encode(prefix_tokens)
return prefix_ids
```

File: src/transformers/models/whisper/tokenization_whisper.py [hunk 5]
```diff
@@ -231,27 +361,57 @@ def bpe(self, token):
...
+        return bos_sequence
 
-        return output + bos_token_ids + token_ids_1
+    @property
+    def prefix_tokens(self):
+        return self.set_bos_sequence(
+            language=self.language,
+            task=self.task,
+            predict_timestamps=self.predict_timestamps,
+        )
```
[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply src/transformers/models/whisper/tokenization_whisper.py]: Yep good point! Resolved in https://github.com/huggingface/transformers/pull/19921/commits/894f842df67e7fae2637eb044ecda5b7f1690a67

File: src/transformers/models/whisper/tokenization_whisper.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply src/transformers/models/whisper/tokenization_whisper.py]: It's probably unlikely someone will load a tokenizer and want to set it to two different languages -> setting the tokenizer language is just for fine-tuning. In most cases, this will involve just one language.

But I'll add the functionality so it's there though as it's few extra lines of code! I think `self.language = ... ` is preferable here to override the language for the tokenizer.

[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply src/transformers/models/whisper/tokenization_whisper.py]: Resolved in https://github.com/huggingface/transformers/pull/19921/commits/e9cb54612d5218365223546c32fa1d36145a6e75

**Comment by sanchit-gandhi** [comment]
I'm not sure if Patrick currently has the bandwidth to review this, @sgugger would you be able to take a look if you've got a spare few minutes? Thanks! 🙏

File: src/transformers/models/whisper/tokenization_whisper.py [hunk 6]
```diff
@@ -109,16 +230,22 @@ class WhisperTokenizer(PreTrainedTokenizer):
...
             The beginning of sequence token.
         eos_token (`str`, *optional*, defaults to `"<|endoftext|>"`):
             The end of sequence token.
         add_prefix_space (`bool`, *optional*, defaults to `False`):
             Whether or not to add an initial space to the input. This allows to treat the leading word just as any
             other word.
-        add_bos_token (`bool`, *optional*, defaults to `False`):
-            Whether or not to add an initial <|endoftext|> to the input. This allows to treat the leading word just as
-            any other word.
+        language (`str`, *optional*, defaults to `None`):
```
Comment (sgugger) [inline src/transformers/models/whisper/tokenization_whisper.py:240]:
```suggestion
        language (`str`, *optional*):
```

File: src/transformers/models/whisper/tokenization_whisper.py [hunk 7]
```diff
@@ -109,16 +230,22 @@ class WhisperTokenizer(PreTrainedTokenizer):
...
+            The language of the transcription text. The corresponding language id token is appended to the start of the
+            sequence for multilingual speech recognition and speech translation tasks, e.g. for Spanish the token
+            `<|es|>` is appended to the start of sequence. This should be used for multilingual fine-tuning only.
```
Comment (sgugger) [inline src/transformers/models/whisper/tokenization_whisper.py:243]:
```suggestion
            `"<|es|>"` is appended to the start of sequence. This should be used for multilingual fine-tuning only.
```

File: src/transformers/models/whisper/tokenization_whisper.py [hunk 8]
```diff
@@ -109,16 +230,22 @@ class WhisperTokenizer(PreTrainedTokenizer):
...
+        task (`str`, *optional*, defaults to `None`):
```
Comment (sgugger) [inline src/transformers/models/whisper/tokenization_whisper.py:244]:
```suggestion
        task (`str`, *optional*):
```

File: src/transformers/models/whisper/tokenization_whisper.py [hunk 9]
```diff
@@ -109,16 +230,22 @@ class WhisperTokenizer(PreTrainedTokenizer):
...
+            Task identifier to append at the start of sequence (if any). This should be used for mulitlingual
+            fine-tuning, with `transcribe` for speech recognition and `translate` for speech translation.
```
Comment (sgugger) [inline src/transformers/models/whisper/tokenization_whisper.py:246]:
```suggestion
            fine-tuning, with `"transcribe"` for speech recognition and `"translate"` for speech translation.
```

File: src/transformers/models/whisper/tokenization_whisper.py [hunk 10]
```diff
@@ -231,27 +362,74 @@ def bpe(self, token):
...
-            bos_token_ids = []
+    def set_prefix_tokens(self, language: str = None, task: str = None, predict_timestamps: bool = None):
+        """
+        Override the prefix tokens appended to the start of the label sequence. This method can be used standalone to
+        update the prefix tokens as required when fine-tuning. Example:
 
-        output = bos_token_ids + token_ids_0
+        ```python
+        >>> tokenizer = WhisperTokenizer.from_pretrained("openai/whisper-tiny", language="spanish")
+        >>> tokenizer.set_prefix_tokens(language="french")  # update the language prefix token
```
Comment (sgugger) [inline src/transformers/models/whisper/tokenization_whisper.py:372]: Are we switching from Spanish to Franch here? Would be useful if the comment was clearer on that.

**Review by sgugger — APPROVED** [review]
Just have some nits on the doc!

File: src/transformers/models/whisper/tokenization_whisper.py (hunk 10 above)
Comment (sanchit-gandhi) [reply src/transformers/models/whisper/tokenization_whisper.py]: Indeed we are! Resolved in: https://github.com/huggingface/transformers/pull/19921/commits/aa8f4cf7c09e60d2ae68b32f58a71213af157c95

**Review by patrickvonplaten — COMMENTED** [review]
Looks good to me, but let's please add one test for `set_prefix_tokens` e.g. changing the language of the tokenizer on the fly

**Comment by sanchit-gandhi** [comment]
Test for `set_prefix_tokens` in https://github.com/huggingface/transformers/pull/19921/commits/e98821f12ba9a899d2907ebcb7d114aff8712c0b

**Comment by patrickvonplaten** [comment]
Cool good to merge for me

File: tests/models/whisper/test_tokenization_whisper.py [hunk 11]
```diff
@@ -176,15 +182,48 @@ def test_tokenizer_decode_ignores_language_codes(self):
...
+            [START_OF_TRANSCRIPT, ES_CODE, TRANSLATE, NOTIMESTAMPS, 17356, 290, 2513, 220,
+             END_OF_TRANSCRIPT, END_OF_TRANSCRIPT, END_OF_TRANSCRIPT],
+            [START_OF_TRANSCRIPT, ES_CODE, TRANSLATE, NOTIMESTAMPS, 17356, 290, 2513, 369,
+             2279, 812, END_OF_TRANSCRIPT]
         ]
         # fmt: on
 
         self.assertListEqual(batch_output, EXPECTED_MULTI)
+
+    def test_set_prefix_tokens(self):
```
Comment (ArthurZucker) [inline tests/models/whisper/test_tokenization_whisper.py:202]: NIce 👍🏻

**Review by ArthurZucker — APPROVED** [review]
LGTM thanks a lot
